### ISSUE TITLE
Guidance to update an ActiveRecord DB adapter to work with 7.1

### ISSUE THREAD

**Issue by mkolmar** [rails/rails#52551]
This is not a bug report. The issue is that ActiveRecord 7.1 seems to have introduced a change that breaks the ActiveRecord ODBC adapter. I am not the original author of the ODBC adapter, which is based on a gem that worked with Rails 5. As such, I am trying to update this adapter to work with Rails 7.1 without a clear understanding of how it works, or the changes from 7.0 to 7.1.

The ActiveRecord ODBC adapter works with Snowflake DB by using Snowflake's ODBC driver. I am not aware of another similar adapter that is maintained.

In its current condition, activerecord-odbc-adapter works reasonably well with 7.0 but not with 7.1. Most likely, this is due to a change, "Introduce a simpler constructor API for ActiveRecord database adapters". Is documentation available that would give some guidance about how to upgrade the adapter to work with 7.1?

From what I can piece together looking at the Postgres adapter and others, the initialize method in ActiveRecord::ConnectionAdapters::ODBCAdapter should change to something like:

```
def initialize(...)
  super
  # other code to map param names or set instance variables?
  # @raw_connection = ??
  # @connection_parameters = ??
end

```
It appears that prior references to `@connection` change to `@raw_connection` in some places in some of the 7.1 adapters. I also see new methods such as `internal_execute()` in 7.1. I am not sure about the extent of the required changes to make this ODBC adapter or other adapter work with 7.1.

### Steps to reproduce

Able to run queries etc. against Snowflake DB using 7.0. After update to Rails 7.1, various references or method calls on `@connection` fail because the value is nil.

```ruby
# Your reproduction script goes here
```

### System configuration
**Rails version**:
7.0.8 or 7.1.3

**Ruby version**:
3.2.3

**Comment by lfv89** [comment]
Could you reproduce it with this template? It will help other people replicate it too:

https://github.com/rails/rails/blob/main/guides/bug_report_templates/active_record.rb

**Comment by mkolmar** [comment]
The database is read-only, so it is not possible to create tables or insert rows. I might be able to create a similar script, but it would have to select from one of the existing tables. (A lot of Snowflake DB users are probably in the same situation.) The problem is easy to reproduce in the rails console.

I strongly suspect the problem begins when `ActiveRecord::ConnectionAdapters::ODBCAdapter.new` is called. It seems 7.1 calls with a config, rather than the parameters as they were through Rails 7.0. Consequently, the adapter does not initialize correctly, although the call does not fail. After that, any DB operation fails. For example, `@connection` is nil here:

```
    # Returns an array of Column objects for the table specified by
    # +table_name+.
    def columns(table_name, _name = nil)
      stmt = @connection.columns(native_case(table_name.to_s))
      ...
    end
```

It does not look like it would be sufficient to change `initialize()` (and `reconnect!` & `disconnect!`), that `@connection` is no longer used. Postgres adapter for example uses `with_raw_connection do ...`, which yields `@raw_connection`. Is that the intended replacement for the method calls on `@connection'?

Is documentation available, or a discussion of the changes that were made to SQL Server or Oracle adapters? I can make some inferences by looking at source code. But the scale of the required changes is not at all clear, so I am at a loss.

The adapter is activerecord-odbc-adapter, on GitHub under my account.

**Comment by mkolmar** [comment]
The incompatibility between the activerecord-odbc-adapter and ActiveRecord 7.1 (and later) is definitely the result of changes in ActiveRecord. I have been unable to find any documentation or discussion threads that cover these changes, or actionable advice.

I was able to make code changes so that initialize() accepts a config hash, and some related changes. But this only seems to address the deprecation in 7.0, and the adapter still breaks with 7.1. Then 7.2 has more changes around connect and reconnect that I have not had a chance to investigate.

Quite frankly, there seems to be little interest in using ODBC or Snowflake DB with Rails. I simply do not have the time and resources to try to update/fix the ODBC adapter, and maintain it at a level that a sane person could use it in a production application ongoing. Further, I cannot put this dependency on the next person, even if I am able to fix it now.

The switch to Snowflake DB (from Oracle) is driven by the vendor. Under the circumstances, I will have to abandon using Rails with Snowflake, as this combination appears to be unsupportable as a practical matter. Snowflake has a REST API, so we should use that and bypass ODBC.

That said, any help or information would be much appreciated.

**Comment by aidanharan** [comment]
There is no specific documentation on how to upgrade ActiveRecord adapters. The Rails source code and pull requests are your best source of information. You can use the Rails core adapters (MySQL/PostgreSQL/SQLite) as a guide on how to upgrade your adapter but your adapter may also contain custom changes.

As reference you can see the PRs required to upgrade the SQL Server adapter from Rails 7.0 to 7.1 at https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pulls?q=is%3Apr+is%3Aclosed+label%3Arails-7.1+sort%3Acreated-asc

Regarding the config hash change in the initializer you mentioned, you can see the relevant changes in the following PRs:

- Rails: https://github.com/rails/rails/pull/44591
- SQL Server adapter: https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/1065

By looks at both you may be able to work out the changes required for your `activerecord-odbc-adapter` gem. 

The internals of ActiveRecord continue to evolve and so adapters need to be continually updated to accomodate these changes. The Rails team only ensure that the core adapters (MySQL/PostgreSQL/SQLite/Trilogy) are kept up-to-date. It is up to the maintainers of other adapters to make the same changes.

**Comment by mkolmar** [comment]
For the benefit of anyone who finds this issue in the future, I was able to leverage https://github.com/rinsed-org/rb-snowflake-client.

I forked the project and have made a few enhancements so far, e.g. support for roles.

Some models in our application now include ActiveModel, rather than using ActiveRecord. A mixin instantiates ActiveModel objects from the array of hashes that the Snowflake client provides. Then we can use simple accessors that return the value of the attribute, and non-trivial methods in other cases. I could make a small gem along this line, if one does not exist already.
